foolbox/attacks/adef_attack.py

```diff
@@ -0,0 +1,372 @@
+        super(ADefAttack, self).__init__(model=model, criterion=criterion)
+        self.vector_field = None
+
+    @call_decorator
+    def __call__(self, input_or_adv, unpack=True, max_iter=100,
+                 max_norm=np.inf, label=None, smooth=1.0, subsample=10):
+
+        """Parameters
+        ----------
+        input_or_adv : `numpy.ndarray` or :class:`Adversarial`
+            The original, unperturbed input as a `numpy.ndarray` or
+            an :class:`Adversarial` instance.
+        label : int
+            The reference label of the original input. Must be passed
+            if `a` is a `numpy.ndarray`, must not be passed if `a` is
+            an :class:`Adversarial` instance.
+        unpack : bool
+            If true, returns the adversarial input, otherwise returns
+            the Adversarial object.
+        max_iter : int > 0
+            Maximum number of iterations (default max_iter = 100).
+        max_norm : float
+            Maximum l2 norm of vector field (default max_norm = numpy.inf).
+        smooth : float >= 0
+            Width of the Gaussian kernel used for smoothing.
+            (default is smooth = 0 for no smoothing).
+        subsample : int
+            Limit on the number of the most likely classes that should
+            be considered. A small value is usually sufficient and much
+            faster.
+
+        """
+        a = input_or_adv
+        del input_or_adv
+        del label
+        del unpack
+
+        if not a.has_gradient():
+            return
+
+        perturbed = a.original_image.copy()  # is updated in every iteration
+
+        # image_original is not updated, but kept as a copy
+        image_original = a.original_image.copy()
+        target_class = a.target_class()
+        targeted = target_class is not None
+        original_label = a.original_class
+
+        # ADef targets classes according to their prediction score. If the
+        # attack is untargeted, ADef will take the labels of the top
+        # 'subsample' classes. This is faster than taking all the remaining
+        # 999 classes of ImageNet into account. For a targeted attack,
+        # it is necessary to find the probability of this class and pass
+        # this index to ind_of_candidates (not the actual target).
+        if targeted is False:
+            # choose the top-k classes
+            # Include the original label in the list of targets (index 0).
+            logging.info('Only testing the top-{} classes'.format(subsample))
+            assert isinstance(subsample, int)
+            ind_of_candidates = np.arange(subsample)
+        else:
+            logits, _ = a.predictions(perturbed)
+            pred_sorted = (-logits).argsort()
+            index_of_target_class, = np.where(pred_sorted == target_class)
+            ind_of_candidates = index_of_target_class
+            # Include the original label in the list of targets (index 0).
```

Comment:
as I said, you cannot assume that the original label is the one with the largest index

Reply:
ok, got rid of that altogether. It didn't seem to make much sense anyway. In the original code, they included the original label index in their list of targets, but then skipped it anyway when looping over possible targets and it seems never used it again. Hm.